test_streamwarp.py: Little-endian buffer not supported on big-endian compiler

## Description

Tests from test_streamwarp.py fail in big endian architectures. [On s390x](https://launchpadlibrarian.net/684548282/buildlog_ubuntu-mantic-s390x.dipy_1.7.0-2_BUILDING.txt.gz):

```
=================================== FAILURES ===================================
_______________________________ test_bundlewarp ________________________________

    def test_bundlewarp():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)

dipy/align/tests/test_streamwarp.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler

dipy/tracking/streamlinespeed.pyx:338: ValueError
_________________________ test_bundlewarp_vector_filed _________________________

    def test_bundlewarp_vector_filed():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)

dipy/align/tests/test_streamwarp.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler

dipy/tracking/streamlinespeed.pyx:338: ValueError
__________________________ test_bundle_shape_profile ___________________________

    def test_bundle_shape_profile():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)

dipy/align/tests/test_streamwarp.py:62: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler

dipy/tracking/streamlinespeed.pyx:338: ValueError
```

I suppose a adding big endian checks are desired, but for what it's worth downstream I proposed the attached patch to simply skip the tests on big endian architectures.
[skip-streamwarp-test-in-big-endian.txt](https://github.com/dipy/dipy/files/12514607/skip-streamwarp-test-in-big-endian.txt)


## Way to reproduce
- [x] Operating system and version: Ubuntu 23.10.
- [x] Python version: 3.11.
- [x] dipy version: 1.7.0
- [x] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * import numpy; 1.24.2
    * import scipy; 1:10.1
    * import nibabel; 5.1.0
    * import h5py; 3.7.0
    * import cvxpy;
    * import fury;

I'm still seeing those endianness errors in 1.8.0:

<details>
<summary>Test failures on `s390x`</summary>

```python
=================================== FAILURES ===================================
_______________________________ test_bundlewarp ________________________________
    def test_bundlewarp():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)
dipy/align/tests/test_streamwarp.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler
streamlinespeed.pyx:338: ValueError
_________________________ test_bundlewarp_vector_filed _________________________
    def test_bundlewarp_vector_filed():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)
dipy/align/tests/test_streamwarp.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler
streamlinespeed.pyx:338: ValueError
__________________________ test_bundle_shape_profile ___________________________
    def test_bundle_shape_profile():
    
        cingulum_bundles = two_cingulum_bundles()
    
        cb1 = Streamlines(cingulum_bundles[0])
>       cb1 = set_number_of_points(cb1, 20)
dipy/align/tests/test_streamwarp.py:62: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   ValueError: Little-endian buffer not supported on big-endian compiler
streamlinespeed.pyx:338: ValueError
______________________________ test_sphere_dtypes ______________________________
    def test_sphere_dtypes():
        for sphere_name, sphere_path in SPHERE_FILES.items():
            sphere_data = np.load(sphere_path)
>           npt.assert_equal(sphere_data['vertices'].dtype, np.float64)
E           AssertionError: 
E           Items are not equal:
E            ACTUAL: dtype('<f8')
E            DESIRED: <class 'numpy.float64'>
dipy/data/tests/test_data.py:8: AssertionError
=============================== warnings summary ===============================
../../BUILDROOT/python-dipy-1.8.0-5.fc41.s390x/usr/lib64/python3.12/site-packages/dipy/viz/__init__.py:22
  /builddir/build/BUILDROOT/python-dipy-1.8.0-5.fc41.s390x/usr/lib64/python3.12/site-packages/dipy/viz/__init__.py:22: UserWarning: You do not have FURY installed. Therefore, 3D visualization functions will not work for you. Please install or upgrade FURY using pip install -U furyFor detailed installation instructions visit: https://fury.gl/
    warnings.warn(
dipy/align/tests/test_reslice.py: 7 warnings
dipy/denoise/tests/test_gibbs.py: 18 warnings
dipy/workflows/tests/test_denoise.py: 1 warning
  /usr/lib64/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=9419) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()
dipy/workflows/tests/test_tracking.py::test_particle_filtering_tracking_workflows
  /builddir/build/BUILD/dipy-1.8.0/dipy/workflows/tests/test_tracking.py:73: RuntimeWarning: invalid value encountered in cast
    save_nifti(path, arr.astype(np.uint8), affine)
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED dipy/align/tests/test_streamwarp.py::test_bundlewarp - ValueError: Lit...
FAILED dipy/align/tests/test_streamwarp.py::test_bundlewarp_vector_filed - Va...
FAILED dipy/align/tests/test_streamwarp.py::test_bundle_shape_profile - Value...
FAILED dipy/data/tests/test_data.py::test_sphere_dtypes - AssertionError: 
= 4 failed, 784 passed, 81 skipped, 16 deselected, 28 warnings in 881.67s (0:14:41) =
```
</details>

Let me know if you prefer a new issue. But this looks pretty much like what OP reported.

--- a/dipy/io/tests/test_stateful_tractogram.py
+++ b/dipy/io/tests/test_stateful_tractogram.py
@@ -1,6 +1,7 @@
 from copy import deepcopy
 import os
 from os.path import join as pjoin
+import sys
 from tempfile import TemporaryDirectory
 from urllib.error import HTTPError, URLError
 
@@ -18,6 +19,7 @@ from dipy.testing.decorators import set_random_number_generator
 from dipy.utils.optpkg import optional_package
 
 fury, have_fury, setup_module = optional_package("fury", min_version="0.10.0")
+is_big_endian = "big" in sys.byteorder.lower()
 
 
 FILEPATH_DIX, POINTS_DATA, STREAMLINES_DATA = None, None, None
@@ -41,6 +43,7 @@ def teardown_module():
     FILEPATH_DIX, POINTS_DATA, STREAMLINES_DATA = None, None, None
 
 
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_direct_trx_loading():
     trx = tmm.load(FILEPATH_DIX["gs.trx"])
     tmp_dir = deepcopy(trx._uncompressed_folder_handle.name)
@@ -74,6 +77,7 @@ def test_tck_equal_in_vox_space():
     assert_allclose(tmp_points_vox, sft.streamlines.get_data(), atol=1e-3, rtol=1e-6)
 
 
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_trx_equal_in_vox_space():
     sft = load_tractogram(
         FILEPATH_DIX["gs.trx"], FILEPATH_DIX["gs.nii"], to_space=Space.VOX
@@ -117,6 +121,7 @@ def test_tck_equal_in_rasmm_space():
     assert_allclose(tmp_points_rasmm, sft.streamlines.get_data(), atol=1e-3, rtol=1e-6)
 
 
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_trx_equal_in_rasmm_space():
     sft = load_tractogram(
         FILEPATH_DIX["gs.trx"], FILEPATH_DIX["gs.nii"], to_space=Space.RASMM
@@ -160,6 +165,7 @@ def test_tck_equal_in_voxmm_space():
     assert_allclose(tmp_points_voxmm, sft.streamlines.get_data(), atol=1e-3, rtol=1e-6)
 
 
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_trx_equal_in_voxmm_space():
     sft = load_tractogram(
         FILEPATH_DIX["gs.trx"], FILEPATH_DIX["gs.nii"], to_space=Space.VOXMM
@@ -366,6 +372,7 @@ def test_tck_iterative_saving_loading():
             save_tractogram(sft_iter, pjoin(tmp_dir, "gs_iter.tck"))
 
 
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_trx_iterative_saving_loading():
     sft = load_tractogram(
         FILEPATH_DIX["gs.trx"], FILEPATH_DIX["gs.nii"], to_space=Space.RASMM

--- a/dipy/segment/tests/test_bundles.py
+++ b/dipy/segment/tests/test_bundles.py
@@ -1,9 +1,7 @@
-import sys
 import warnings
 
 import numpy as np
 from numpy.testing import assert_almost_equal, assert_equal
-import pytest
 
 from dipy.data import get_fnames
 from dipy.io.streamline import load_tractogram
@@ -12,8 +10,6 @@ from dipy.segment.clustering import qbx_and_merge
 from dipy.testing.decorators import set_random_number_generator
 from dipy.tracking.distances import bundles_distances_mam
 from dipy.tracking.streamline import Streamlines
-
-is_big_endian = "big" in sys.byteorder.lower()
 
 
 def setup_module():
@@ -35,7 +31,6 @@ def setup_module():
     f.extend(f3)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_check_defaults():
     rb = RecoBundles(f, greater_than=0, clust_thr=10)
 
@@ -69,7 +64,6 @@ def test_rb_check_defaults():
         assert_equal(row.min(), 0)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_disable_slr():
     rb = RecoBundles(f, greater_than=0, clust_thr=10)
 
@@ -103,7 +97,6 @@ def test_rb_disable_slr():
         assert_equal(row.min(), 0)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 @set_random_number_generator(42)
 def test_rb_slr_threads(rng):
     rb_multi = RecoBundles(f, greater_than=0, clust_thr=10, rng=rng)
@@ -133,7 +126,6 @@ def test_rb_slr_threads(rng):
         assert_almost_equal(row.min(), 0, decimal=4)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_no_verbose_and_mam():
     rb = RecoBundles(f, greater_than=0, clust_thr=10, verbose=False)
 
@@ -171,7 +163,6 @@ def test_rb_no_verbose_and_mam():
         assert_equal(row.min(), 0)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_clustermap():
     cluster_map = qbx_and_merge(f, thresholds=[40, 25, 20, 10])
 
@@ -208,7 +199,6 @@ def test_rb_clustermap():
         assert_equal(row.min(), 0)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_no_neighb():
     # what if no neighbors are found? No recognition
 
@@ -245,7 +235,6 @@ def test_rb_no_neighb():
         assert_equal(len(rec_trans), 0)
 
 
-@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_rb_reduction_mam():
     rb = RecoBundles(f, greater_than=0, clust_thr=10, verbose=True)
 

--- a/dipy/tracking/streamlinespeed.pyx
+++ b/dipy/tracking/streamlinespeed.pyx
@@ -7,7 +7,7 @@ from libc.stdlib cimport malloc, free
 cimport numpy as cnp
 
 from dipy.tracking import Streamlines
-
+from dipy.utils.arrfuncs import as_native_array
 
 cdef extern from "dpy_math.h" nogil:
     bint dpy_isnan(double x)
@@ -99,15 +99,15 @@ def length(streamlines):
 
         arclengths = np.zeros(len(streamlines), dtype=np.float64)
 
-        if streamlines._data.dtype == np.float32:
+        if as_native_array(streamlines._data).dtype == np.float32:
             c_arclengths_from_arraysequence[float2d](
-                                    streamlines._data,
+                                    as_native_array(streamlines._data),
                                     streamlines._offsets.astype(np.intp),
                                     streamlines._lengths.astype(np.intp),
                                     arclengths)
         else:
             c_arclengths_from_arraysequence[double2d](
-                                      streamlines._data,
+                                      as_native_array(streamlines._data),
                                       streamlines._offsets.astype(np.intp),
                                       streamlines._lengths.astype(np.intp),
                                       arclengths)
@@ -319,7 +319,7 @@ def set_number_of_points(streamlines, nb_points=3):
             return Streamlines()
 
         nb_streamlines = len(streamlines)
-        dtype = streamlines._data.dtype
+        dtype = as_native_array(streamlines._data).dtype
         new_streamlines = Streamlines()
         new_streamlines._data = np.zeros((nb_streamlines * nb_points, 3),
                                          dtype=dtype)
@@ -330,12 +330,12 @@ def set_number_of_points(streamlines, nb_points=3):
 
         if dtype == np.float32:
             c_set_number_of_points_from_arraysequence[float2d](
-                streamlines._data, streamlines._offsets.astype(np.intp),
+                as_native_array(streamlines._data), streamlines._offsets.astype(np.intp),
                 streamlines._lengths.astype(np.intp), nb_points,
                 new_streamlines._data)
         else:
             c_set_number_of_points_from_arraysequence[double2d](
-                streamlines._data, streamlines._offsets.astype(np.intp),
+                as_native_array(streamlines._data), streamlines._offsets.astype(np.intp),
                 streamlines._lengths.astype(np.intp), nb_points,
                 new_streamlines._data)
 

--- a/dipy/utils/tests/test_tractogram.py
+++ b/dipy/utils/tests/test_tractogram.py
@@ -1,10 +1,16 @@
+import sys
+
+import pytest
 import trx.trx_file_memmap as tmm
 
 from dipy.data import get_fnames
 from dipy.io.streamline import load_tractogram
 from dipy.utils.tractogram import concatenate_tractogram
 
+is_big_endian = "big" in sys.byteorder.lower()
 
+
+@pytest.mark.skipif(is_big_endian, reason="Little Endian architecture required")
 def test_concatenate():
     filepath_dix, _, _ = get_fnames(name="gold_standard_tracks")
     sft = load_tractogram(filepath_dix["gs.trk"], filepath_dix["gs.nii"])

--- a/dipy/workflows/tests/test_io.py
+++ b/dipy/workflows/tests/test_io.py
@@ -36,6 +36,8 @@ logging.basicConfig(
     filemode="w",
 )
 
+is_big_endian = "big" in sys.byteorder.lower()
+
 
 def test_io_info():
     fimg, fbvals, fbvecs = get_fnames(name="small_101D")
@@ -50,8 +52,12 @@ def test_io_info():
     io_info_flow.run([fimg, fbvals, fvecs], b0_threshold=20, bvecs_tol=0.001)
 
     filepath_dix, _, _ = get_fnames(name="gold_standard_tracks")
+    if not is_big_endian:
+        io_info_flow = IoInfoFlow()
+        io_info_flow.run(filepath_dix["gs.trx"])
+
     io_info_flow = IoInfoFlow()
-    io_info_flow.run([filepath_dix["gs.trx"], filepath_dix["gs.trk"]])
+    io_info_flow.run(filepath_dix["gs.trk"])
 
     io_info_flow = IoInfoFlow()
     npt.assert_raises(TypeError, io_info_flow.run, filepath_dix["gs.tck"])
@@ -175,7 +181,14 @@ def test_convert_sh_flow():
 def test_convert_tractogram_flow():
     with TemporaryDirectory() as out_dir:
         data_path, _, _ = get_fnames(name="gold_standard_tracks")
-        input_files = [v for k, v in data_path.items() if k in ["gs.tck", "gs.trx"]]
+        input_files = [
+            v
+            for k, v in data_path.items()
+            if k
+            in [
+                "gs.tck",
+            ]
+        ]
 
         convert_tractogram_flow = ConvertTractogramFlow(mix_names=True)
         convert_tractogram_flow.run(
@@ -187,13 +200,14 @@ def test_convert_tractogram_flow():
             ValueError, convert_tractogram_flow.run, input_files, out_dir=out_dir
         )
 
-        npt.assert_warns(
-            UserWarning,
-            convert_tractogram_flow.run,
-            data_path["gs.trx"],
-            out_dir=out_dir,
-            out_tractogram="gs_converted.trx",
-        )
+        if not is_big_endian:
+            npt.assert_warns(
+                UserWarning,
+                convert_tractogram_flow.run,
+                data_path["gs.trx"],
+                out_dir=out_dir,
+                out_tractogram="gs_converted.trx",
+            )
 
 
 def test_convert_tensors_flow():

